Plugins/GitUIPluginInterfaces/GitRefsFlags.cs

```diff
@@ -1,18 +1,21 @@
-﻿namespace GitCommands.Git.Commands
+﻿namespace GitUIPluginInterfaces
 {
     /// <summary>
     /// Enums requestable in GitRefs() (multiple names can be appended)
     /// Compare to <see ref="GitRefType"/> for actual values of parsed GitRefs
     /// </summary>
-    public enum GetRefsEnum
+    public enum GitRefsFlags
     {
         None = 0,
 
-        Branches = 1 << 0,
+        // Local Branches
+        Heads = 1 << 0,
+
+        // Remote branches
         Remotes = 1 << 1,
         Tags = 1 << 2,
 
-        // All refs, including those abve but also (at least) stash, notes and bisect
+        // All refs, including those above but also (at least) stash, notes and bisect
         All = 1 << 3
```

Comment:
```suggestion
        Everything = 1 << 3 | Heads | Remotes | Tags
```
...or rewrite the enum as
```cs
 Heads      = 0x0001,
 Remotes    = 0x0010,
 Tags       = 0x0100,
 Everything = 0x1111
```

Reply:
Prefer not to change this, then a separate flag is needed for All anyway.
The reason is that HasFlagButNotAll() need to be able to check if All bit is set, then the arguments for git-for-each-ref should be empty and that check cannot be done.

Sure, the most significant bit for All can be dynamically calculated in GetRefsCmd() too.
So the following (and another minor change) is required.
```c#
            static bool HasFlagButNotAll(GitRefs requested, GitRefs flag)
            {
                const GitRefs allOnly = GitRefs.All & ~(GitRefs.Heads | GitRefs.Remotes | GitRefs.Tags);
                return (requested & allOnly) == 0 && (requested & flag) != 0;
            }
```

Reply:
I refactored GetRefsCmd() some to make this handling clearer.

Reply:
🤦‍♂️ Looks like I've made a clown of myself. Here's what I meant to write above:
```cs
[Flags]
public enum GitRefs : byte
{
	None       = 0,
	Heads      = 0x01,          // 00000001
	Remotes    = 0x02,          // 00000010
	Tags       = 0x04,          // 00000100
	Everything = byte.MaxValue  // 11111111
}
```


My point was that `All` makes no sense as a single bit. E.g. `GitRefs.Remotes | GitRefs.Tags` means "_remotes and tags_", but `GitRefs.All | GitRefs.Remotes | GitRefs.Tags` means "_everything_" with "remotes and tags" being totally superficial.

Since the members in the enum aren't representing the exhaustive set of values (which we may consider adding in the future), `All` (or `Everything`) must essentially be represented by "all bits set" value. And to assert whether it is set we just need to read the top bit, e.g. `((int)requested >> 7) == 1`

Reply:
In this scenario, `Other` will not get any data at all or be the same as `Everything` which is confusing for an API usage.
The enums is the API, should be clear from a usage point of view.
This is also why I do not want to have a `None` value. I that value is used alone, the Git command should throw as it is meaningless, and how should `None|Tags` be interpreted?

All is a special mode comparable to Heads/Remotes/Tags even if All contains the latter. So `All|Tags` is the same as `All` regardless if all bits are set or not. Sure, that All is not used with other could be asserted, but that will be a runtime check without any real meaning.

I can add all bits to All but the internal check is a little weird in my opinion, the check must mask the enum to isolate the All bit.

Personally I find 1>>1, 1>>2 etc be easier to understand than listing hexvalues. (binary values work well here, harder when there are many values).

Reply:
`GetRefs` uses this argument to control the filtering, where `All` means `Unfiltered`. Hence the enum value should be defined as `Unfiltered = 0`.

Reply:
> `GetRefs` uses this argument to control the filtering, where `All` means `Unfiltered`. Hence the enum value should be defined as `Unfiltered = 0`.

Good suggestion, that should work
The name of the type should then be `GitRefsFilter`

Reply:
`Unfiltered|Tags` will be the same as `Tags` which could be confusing, that is API documentation.

Reply:
> `Unfiltered|Tags` will be the same as `Tags` which could be confusing, that is API documentation.

Then `GitRefsFilter.None = 0`.

Reply:
Thank you @mstv - this makes sense :)